receiver/rabbitmqreceiver/scraper.go

```diff
@@ -0,0 +1,157 @@
+var messageStatMetrics = []string{
+	deliverStat,
+	publishStat,
+	ackStat,
+	dropUnroutableStat,
+}
+
+// rabbitmqScraper handles scraping of RabbitMQ metrics
+type rabbitmqScraper struct {
+	client         client
+	logger         *zap.Logger
+	cfg            *Config
+	metricsBuilder *metadata.MetricsBuilder
+}
+
+// newScraper creates a new scraper
+func newScraper(logger *zap.Logger, cfg *Config) *rabbitmqScraper {
+	return &rabbitmqScraper{
+		logger:         logger,
+		cfg:            cfg,
+		metricsBuilder: metadata.NewMetricsBuilder(cfg.Metrics),
+	}
+}
+
+// start starts the scraper by creating a new HTTP Client on the scraper
+func (r *rabbitmqScraper) start(ctx context.Context, host component.Host) (err error) {
+	r.client, err = newClient(r.cfg, host, r.logger)
+	return
+}
+
+// scrape collects metrics from the RabbitMQ API
+func (r *rabbitmqScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	metrics := pdata.NewMetrics()
+	now := pdata.NewTimestampFromTime(time.Now())
+	rms := metrics.ResourceMetrics()
+
+	// Validate we don't attempt to scrape without initializing the client
+	if r.client == nil {
+		return metrics, errClientNotInit
+	}
+
+	// Get queues for processing
+	queues, err := r.client.GetQueues(ctx)
+	if err != nil {
+		return metrics, err
+	}
+
+	// Collect metrics for each queue
+	for _, queue := range queues {
+
+		r.collectQueue(queue, now, rms)
+	}
+
+	return metrics, nil
+}
+
+// collectQueue collects metrics
+func (r *rabbitmqScraper) collectQueue(queue *models.Queue, now pdata.Timestamp, rms pdata.ResourceMetricsSlice) {
+	resourceMetric := rms.AppendEmpty()
+	resourceAttrs := resourceMetric.Resource().Attributes()
+	resourceAttrs.InsertString(metadata.A.RabbitmqQueueName, queue.Name)
+	resourceAttrs.InsertString(metadata.A.RabbitmqNodeName, queue.Node)
+	resourceAttrs.InsertString(metadata.A.RabbitmqVhostName, queue.VHost)
+
+	ilms := resourceMetric.InstrumentationLibraryMetrics().AppendEmpty()
+	ilms.InstrumentationLibrary().SetName(instrumentationLibraryName)
+
+	r.metricsBuilder.RecordRabbitmqConsumerCountDataPoint(now, queue.Consumers)
+	r.metricsBuilder.RecordRabbitmqMessageCurrentDataPoint(now, queue.UnacknowledgedMessages, metadata.AttributeMessageState.Unacknowledged)
+	r.metricsBuilder.RecordRabbitmqMessageCurrentDataPoint(now, queue.ReadyMessages, metadata.AttributeMessageState.Ready)
+
+	for _, messageStatMetric := range messageStatMetrics {
+		// Get metric value
+		val, ok := queue.MessageStats[messageStatMetric]
+		// A metric may not exist if the actions that increment it do not occur
+		if !ok {
+			r.logger.Debug("metric not found", zap.String("Metric", messageStatMetric), zap.String("Queue", queue.Name))
+			continue
+		}
+
+		// Convert to int64
+		val64, ok := convertValToInt64(val)
+		if !ok {
+			// Log warning if the metric is not in the format we expect
+			r.logger.Warn("metric not int64", zap.String("Metric", messageStatMetric), zap.String("Queue", queue.Name))
```

Comment:
In cases where an error has occurred during collection, but we can continue on, we need to use the `scrapererror` package to capture partial errors.

Reply:
I see the comment on `convertValToInt64`, about this being an unexpected error. Given there are no other partial failure mechanisms, if you think it's extremely unlikely then we can maybe skip using partial error. Wouldn't be a bad idea to do it anyways, but I'll leave it up to your judgement.

Reply:
Thanks for the info on `scrapererror` I wasn't aware of it. I think I'm going to keep this as a warning log only for now as the API would need to change in order for it to fail.